utils.go

```diff
@@ -87,13 +88,16 @@ func getRootlessRuntimeDir(rootlessUID int) (string, error) {
 	if err == nil {
 		return runtimeDir, nil
 	}
-	tmpDir := fmt.Sprintf("/run/user/%d", rootlessUID)
-	st, err := system.Stat(tmpDir)
-	if err == nil && int(st.UID()) == os.Getuid() && st.Mode()&0700 == 0700 && st.Mode()&0066 == 0000 {
-		return tmpDir, nil
+	initCommand, err := ioutil.ReadFile("/proc/1/comm")
+	if err != nil || string(initCommand) == "systemd" {
+		tmpDir := fmt.Sprintf("/run/user/%d", rootlessUID)
+		st, err := system.Lstat(tmpDir)
+		if err == nil && int(st.UID()) == os.Getuid() && st.Mode()&0700 == 0700 && st.Mode()&0066 == 0000 {
+			return tmpDir, nil
+		}
 	}
-	tmpDir = fmt.Sprintf("%s/%d", os.TempDir(), rootlessUID)
-	if err := os.MkdirAll(tmpDir, 0700); err != nil {
+	tmpDir := fmt.Sprintf("%s/containers-user-%d", os.TempDir(), rootlessUID)
+	if err := os.Mkdir(tmpDir, 0700); err != nil && !os.IsExist(err) {
```

Comment:
Still https://github.com/containers/storage/pull/568#discussion_r400156458 . This is no better than `MkdirAll`.

I’m increasingly convinced this does require 100% code coverage and attack scenario simulation — except that the previous version is also broken.

Reply:
How about if we check if the directory exists, and proper permissions, we use it.
If not then we create a random tmporary directory ioutil.TempDir() https://golang.org/pkg/io/ioutil/#TempDir
Then do a os.Rename(newTmpDir,tmpDir)
Does that solve the race condition.

Reply:
@mtrmac WYT?

Reply:
That would probably resolve the race condition (write tests!) but I don’t think it would work, in `S_ISVTX` `/tmp` the user is not allowed to remove another users’ files.

Back to https://github.com/containers/storage/pull/568#discussion_r394967980 .

(To be clear, it would clearly be more safe than the current `MkdirAll` or the current PR. Better DOS than a privilege escalation.)

Reply:
RIght it would fail, which would tell the user he is under attack.
If I saw this happen, I would see permission denied on /tmp/container-user-3267, and then I would check what user created this file with my UID, and I would have a good idea something suspicions is going on.

Reply:
If that is OK, the code could just as well raise an alarm without trying to create the temporary file and `Rename`.

(As for whether such a behavior in `/tmp`, exposing the non-Systemd user to a trivial DOS, is desirable: as established elsewhere in the PR, I’m not spending time on making it work, only on not making it insecure.)

Reply:
I believe this would be secure and would satisfy the need.  These issues on creating content in /tmp in well known locations have always existed.  We just need to make sure we don't make a mistake in a potentially risky environment.

Reply:
@mtrmac You mean that other user created `/tmp/folder` with 777 permissions and watching what content will appear in this folder. I can't understand why you are talking about this issue here. You can grep all storage project and find 10-20 `!os.IsExist(err)`. Please create new issue and we can discuss here what to do with `!os.IsExist(err)`.

Reply:
@mtrmac, BTW I see no difference between `/tmp` and `/run/user`. Another user can fool software and redirect system calls from `/proc/1/comm` to fake file and create symlink from `/tmp` to `/run/user`. I am sure that solution for `os.IsExist(err)` shouldn't depend on any parameter like initialization system detection.

Reply:
@mtrmac We can check whether target folder was created by previous `Mkdir` call. It is possible to verify permissions after `os.IsExist`. We can create project scope utility and use it instead of `Mkdir` everywhere. What do you think about it?

Reply:
> @mtrmac You mean that other user created `/tmp/folder` with 777 permissions and watching what content will appear in this folder. I can't understand why you are talking about this issue here. You can grep all storage project and find 10-20 `!os.IsExist(err)`. 

I expect _most_ of them not to happen in world-writable paths (but, full disclosure, I didn’t check).

Still, sure, at least the original `MkdirAll` here was insecure, and I _somewhat_ agree that however that code was introduced, it should be fixed the same way. But then this PR wants to _change_ the behavior away from `/run/user` towards that path on more systems, and that requires that path to, at least, not be known unsafe.

Reply:
> @mtrmac, BTW I see no difference between `/tmp` and `/run/user`. Another user can fool software and redirect system calls from `/proc/1/comm` to fake file and create symlink from `/tmp` to `/run/user`.

How exactly? Root user attacking other users, VM creators attacking contents of VMs, users attacking themselves, and similar things are not in the threat model I care about, at least.